### PR TITLE
Switched to use leo_net:chunked_send

### PR DESCRIPTION
**Pull request by windkit**
### Problem Description
Current Implementation of sending object to client would catch a problem if the transfer rate is too low.
in `leo_gateway` a chunk is sent at once with `ranch_tcp:send/2` has a send timeout option which defaults to 30 seconds. In other words, if the chunk could not be sent within the timeout, the transfer would fail.

With default value of chunks size (5MB) and timeout (30 seconds), if the transfer rate is lower than 167KB/s, the send would fail with timeout.

### Proposed Solution
As a result, this PR (together with the PR in `leo_commons`) proposes to add a layer to break down object transfer to smaller chunks. 

The sending chunk size is configurable in `leo_gateway.conf` as `http.sending_chunk_len = 131072`

The default value is 128KB, implying a 4KB/s lower bound.

### Related Library PR
https://github.com/leo-project/leo_commons/pull/6

### Related Issue
https://github.com/leo-project/leofs/issues/429

### CODE REVIEW

File: include/leo_http.hrl [hunk 1]
```diff
@@ -390,6 +392,7 @@
           custom_header_settings     :: list() | undefined,     %% http custom header settings
           timeout_for_header         :: pos_integer(),          %% Timeout for reading header
           timeout_for_body           :: pos_integer(),          %% Timeout for reading body
+          sending_chunk_len = 0      :: pos_integer(),          %% sending chunk length
```
Comment (mocchira) [inline include/leo_http.hrl]: This new parameter is the send version corresponding to the existing parameter 'reading_chunked_obj_len' the read version.
So it'd be good to comply with the existing naming rule for LeoFS users.

s/sending_chunk_len/sending_chunked_obj_len/

File: priv/leo_gateway.conf [hunk 2]
```diff
@@ -79,6 +79,9 @@ http.max_keepalive = 4096
 ## HTTP timeout for reading body
 ## http.timeout_for_body = 15000
 
+## HTTP sending chunk length
+## http.sending_chunk_len = 131072
+
```
Comment (mocchira) [inline priv/leo_gateway.conf]: The default should NOT be affected to existing user environments.

s/131072/5242880/

File: priv/leo_gateway.schema [hunk 3]
```diff
@@ -228,6 +228,15 @@
...
  ]}.
 
+%% @doc Timeout for sending chunk length
+{mapping,
+ "http.sending_chunk_len",
+ "leo_gateway.http.sending_chunk_len",
+ [
+  {datatype, integer},
+  {default, 131072}
+ ]}.
```
Comment (mocchira) [inline priv/leo_gateway.schema]: Same with the above comment.

**Comment by mocchira** [comment]
@windkit Please modify leo_common version to reference the new one including leo_net

**Comment by mocchira** [comment]
@windkit as you may hear from @yosukehara ,
Please benchmark before/after with various chunk settings to confirm whether there is no performance degradation.

**Comment by mocchira** [comment]
@windkit 
To make sure, please build and test on R16B03-1.
Since I'm not sure `inet:socket` can be used on the above version.

**Comment by windkit** [comment]
Benchmark results are uploaded at
https://github.com/leo-project/notes/tree/master/leofs/benchmark/leofs/1.2/chunked_send/20151029_read

I tested with 128KB/512KB/5MB fixed objects with 3 settings
- original
- chunked_send 128KB chunks
- chunked_send 4KB chunks

Using 128KB chunks shows similar performance with original design, 4KB chunks show ~10% drop for 128KB objects

**Comment by yosukehara** [comment]
@windkit Thanks for sharing the benchmark report. Do you have ideas of improving its performance?

**Comment by windkit** [comment]
@yosukehara I think the drop for 4KB

**Comment by mocchira** [comment]
@yosukehara I think it's a reasonable result to me.
In case of 4KB chunks, 10% drop maybe come from increasing the number of system calls or
decreasing the effective rate of network transfer speed.
So it's reasonable.

**Comment by yosukehara** [comment]
@mocchira ok, I'm with you. Thanks for your comment.

**Comment by windkit** [comment]
@mocchira I have made some changes, please help to review again :)

**Comment by mocchira** [comment]
@windkit Files suffixed by "~" are committed. Please revert those files.
Others except that looks good.

**Comment by windkit** [comment]
@mocchira oops, I have removed them now.

**Comment by mocchira** [comment]
LGTM.
@yosukehara I'll merge this PR if you are OK.

**Comment by yosukehara** [comment]
@mocchira I've just checked this request.  After I released LeoFS v1.4.0-pre2, it will be merged into the develop-branch.

**Comment by mocchira** [comment]
@yosukehara got it.

**Comment by yosukehara** [comment]
@windkit @mocchira Thank you for your contribution. It has been just merged into the develop-branch.

**Pull request merged**